File: mobile_shop_app/lib/features/customer/models/customer.dart
class Customer {
  final int id;
  final int shopId;
  final String name;
  final String mobile;
  final String? alternateMobile;
  final String? email;
  final String? address;
  final String? city;
  final String? notes;
  final String? createdAt;
  final String? updatedAt;

  Customer({
    required this.id,
    required this.shopId,
    required this.name,
    required this.mobile,
    this.alternateMobile,
    this.email,
    this.address,
    this.city,
    this.notes,
    this.createdAt,
    this.updatedAt,
  });

  factory Customer.fromJson(Map<String, dynamic> json) {
    return Customer(
      id: json['id'] is int ? json['id'] : int.parse(json['id'].toString()),
      shopId: json['shop_id'] is int ? json['shop_id'] : int.parse(json['shop_id'].toString()),
      name: json['name'] ?? '',
      mobile: json['mobile'] ?? '',
      alternateMobile: json['alternate_mobile'],
      email: json['email'],
      address: json['address'],
      city: json['city'],
      notes: json['notes'],
      createdAt: json['created_at'],
      updatedAt: json['updated_at'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'shop_id': shopId,
      'name': name,
      'mobile': mobile,
      'alternate_mobile': alternateMobile,
      'email': email,
      'address': address,
      'city': city,
      'notes': notes,
      'created_at': createdAt,
      'updated_at': updatedAt,
    };
  }
}

File: app/Models/Customer.php
<?php

namespace App\Models;

use App\Traits\BelongsToShop;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class Customer extends Model
{
    use SoftDeletes;
    use BelongsToShop, HasFactory;

    protected $fillable = [
        'shop_id',
        'name',
        'mobile',
        'alternate_mobile',
        'email',
        'address',
        'city',
        'notes',
    ];

    /**
     * Get the shop that owns the customer.
     */
    public function shop(): BelongsTo
    {
        return $this->belongsTo(Shop::class);
    }

    /**
     * Get the devices belonging to the customer.
     */
    public function devices(): \Illuminate\Database\Eloquent\Relations\HasMany
    {
        return $this->hasMany(Device::class);
    }

    /**
     * Get the sales belonging to the customer.
     */
    public function sales(): \Illuminate\Database\Eloquent\Relations\HasMany
    {
        return $this->hasMany(Sale::class);
    }

    /**
     * Get the repairs belonging to the customer.
     */
    public function repairs(): \Illuminate\Database\Eloquent\Relations\HasMany
    {
        return $this->hasMany(Repair::class);
    }

    /**
     * Get the warranties belonging to the customer.
     */
    public function warranties(): \Illuminate\Database\Eloquent\Relations\HasMany
    {
        return $this->hasMany(Warranty::class)->whereNull('warranties.deleted_at');
    }

    protected static function booted(): void
    {
        static::deleting(function ($customer) {
            foreach ($customer->warranties()->get() as $warranty) {
                $warranty->delete();
            }
        });
    }
}